xcm/xcm-executor/src/lib.rs

```diff
@@ -128,8 +124,12 @@ impl<Config: config::Config> ExecuteXcm<Config::Call> for XcmExecutor<Config> {
 
 		let mut vm = Self::new(origin);
 
+		// We hash the message here instead of inside the loop, because `message` may have been
+		// reassigned due to the existence of the error handler or the appendix.
+		let message_hash = message.using_encoded(sp_io::hashing::blake2_256);
```

Comment:
The XCM version on the receiving side may not be the same as it was on the sending side (owing to automated translation between versions).

Since all XCMs are always sent as a `VersionedXcm` while on the wire, it's better to hash the encoded `VersionedXcm`. This also has the advantage that it avoids a needless `.encode()`. This ensure that it's definitely the same hash as on the sending side.